storage/raft_test.go

```diff
@@ -35,66 +102,105 @@ func internalChangeReplicasArgs(newNodeID proto.NodeID, newStoreID proto.StoreID
+		mtc.rangeID, mtc.store.StoreID())
+	if err := mtc.rng.AddCmd(proto.InternalChangeReplicas, args, resp, true); err != nil {
 		t.Fatal(err)
 	}
-	store2.db = client.NewKV(&testSender{store: store2}, nil)
-	if err := store2.Start(); err != nil {
+
+	// Verify that the same data is available on the replica.
+	if err := util.IsTrueWithin(func() bool {
+		getArgs, getResp := getArgs([]byte("a"), mtc.rangeID, mtc.stores[1].StoreID())
+		if err := mtc.stores[1].ExecuteCmd(proto.Get, getArgs, getResp); err != nil {
+			return false
+		}
+		return getResp.Value.GetInteger() == 5
+	}, 1*time.Second); err != nil {
 		t.Fatal(err)
 	}
-	defer store2.Stop()
+}
 
-	// Issue a command on the first node before replicating.
-	incArgs, incResp := incrementArgs([]byte("a"), 5, rangeID, tc.store.StoreID())
-	if err := tc.rng.AddCmd(proto.Increment, incArgs, incResp, true); err != nil {
+// TestRestoreReplicas ensures that consensus group membership is properly
+// persisted to disk and restored when a node is stopped and restarted.
+func TestRestoreReplicas(t *testing.T) {
+	mtc := multiTestContext{}
+	mtc.Start(t, 2)
+	defer mtc.Stop()
+
+	args, resp := internalChangeReplicasArgs(mtc.stores[1].Ident, proto.ADD_REPLICA,
+		mtc.rangeID, mtc.store.StoreID())
+	if err := mtc.stores[0].ExecuteCmd(proto.InternalChangeReplicas, args, resp); err != nil {
 		t.Fatal(err)
 	}
 
-	args, resp := internalChangeReplicasArgs(nodeID2, storeID2, proto.ADD_REPLICA,
-		rangeID, tc.store.StoreID())
-	if err := tc.rng.AddCmd(proto.InternalChangeReplicas, args, resp, true); err != nil {
+	// Let the range propagate to the second replica.
+	// TODO(bdarnell): we need an observable indication that this has completed.
+	// TODO(bdarnell): initial creation and replication needs to be atomic;
+	// cutting off the process too soon currently results in a corrupted range.
+	// We need 500ms to be safe in -race tests.
+	time.Sleep(500 * time.Millisecond)
+
+	mtc.StopStore(0)
+	mtc.StopStore(1)
+	mtc.RestartStore(0, t)
+	mtc.RestartStore(1, t)
+
+	// Send a command on each store. The follower will forward to the leader and both
+	// commands will eventually commit.
+	incArgs, incResp := incrementArgs([]byte("a"), 5, mtc.rangeID, mtc.stores[0].StoreID())
+	if err := mtc.stores[0].ExecuteCmd(proto.Increment, incArgs, incResp); err != nil {
+		t.Fatal(err)
+	}
+	incArgs, incResp = incrementArgs([]byte("a"), 11, mtc.rangeID, mtc.stores[1].StoreID())
+	if err := mtc.stores[1].ExecuteCmd(proto.Increment, incArgs, incResp); err != nil {
 		t.Fatal(err)
 	}
 
-	// Verify that the same data is available on the replica.
 	if err := util.IsTrueWithin(func() bool {
-		getArgs, getResp := getArgs([]byte("a"), rangeID, store2.StoreID())
-		if err := store2.ExecuteCmd(proto.Get, getArgs, getResp); err != nil {
+		getArgs, getResp := getArgs([]byte("a"), mtc.rangeID, mtc.stores[1].StoreID())
+		if err := mtc.stores[1].ExecuteCmd(proto.Get, getArgs, getResp); err != nil {
```

Comment:
Unless stores[1] is the leader, this only works because we're not yet returning the truth from Range.IsLeader().